File: KnightsTour/tour.java
// Toufic Hamadeh
// UCID 31234305
// 20-FEB-2013

// Professor Kapleau
// CS-280-002
// Knight's Tour in Java

public class tour {
  public static void main(String[] args) {
    KnightsTour kt = new KnightsTour();
    kt.begin(0, 0);
  }
}

class KnightsTour {
    int chessBoard[][] = new int[8][8];
    int checkBoard[][] = new int[8][8];
    int[] horizontal = {2, 1, -1, -2, -2, -1, 1, 2};
    int[] vertical =   {-1, -2, -2, -1, 1, 2, 2, 1};
    int cRow;
    int cCol;
    int moveNumber;
    int counter = 0;
    int hori, vert;
    int possible = 0;
    int fMove, lMove;
    int tempV, tempH;
       
public void begin(int a, int b){
    for(int i = 0; i < chessBoard.length; i++){
        for(int u = 0; u < chessBoard[i].length; u++){
            chessBoard[i][u] = 0;
        }
    }
    runAccess();
               
    cRow = a;
    cCol = b;
    move();
}
       
public void runAccess(){
    for(int i = 0; i < checkBoard.length; i++){
        for(int u = 0; u < checkBoard[i].length; u++){
            cRow = i;
            cCol = u;
            access();
            checkBoard[i][u] = possible;
        }
    }
}

// Moves the knight
public void move(){
    chessBoard[cRow][cCol] = counter + 1;
    tempH = cCol;
    tempV = cRow;
    runAccess();

    cCol = tempH;
    cRow = tempV;
    access();

    moveNumber = fMove;
    hori = horizontal[moveNumber];
    vert = vertical[moveNumber];
    cRow += vert;
    cCol += hori;
    
    ++counter;
    if(counter == 64){
        runAccess();
        finalPrint();
    } else {
        move();
    }
}

public void finalPrint(){
    System.out.println("\n Knight's Tour\n");
    for(int i = 0; i < chessBoard.length; i++){
        for(int u = 0; u < chessBoard[i].length; u++){
            System.out.printf(" %2d", chessBoard[u][i]);
        }
        System.out.println("\n");
    }
}

public void access(){
               
    moveNumber = 0;
    possible = 0;
    fMove = 0;
    lMove = 9;

    for(int i = 0; i < 8; i++){
        moveNumber = i;
        hori = horizontal[moveNumber];
        vert = vertical[moveNumber];
        if(cRow + vert < checkBoard.length && cCol + hori < checkBoard[i].length && cRow + vert >= 0 && cCol + hori >= 0){
            if(chessBoard[cRow + vert][cCol + hori] == 0){
                ++possible;
                if(checkBoard[cRow + vert][cCol + hori] < lMove){
                    fMove = i;
                    lMove = checkBoard[cRow + vert][cCol + hori];
                }
            }      
        }
    }
}

} // end KnightsTour

File: C++ Projects/Knight's Tour/tour.cpp
// Toufic Hamadeh
// UCID 31234305
// 20-FEB-2013

// Professor Kapleau
// CS-280-002
// Knight's Tour in C++

#include <iostream>
#include <iomanip>
using namespace std;

// Global
const int SIZE = 8;
int chessBoard[SIZE][SIZE];
int start = 1;
int border = 8;
int x, y;

// Function Prototypes
void calculate(int, int);
void tourBoard(int, int);
void printBoard();

// Main
int main(int argc, char *argv[]) {
    tourBoard(0, 0);
    printBoard();

    return 0;
}

void calculate(int i, int j) {
    int a;
    int b;
    int counter;

    a = i;
    b = j;
    counter = 0;

    if ((a >= 0 && b >= 0) && (a < 8 && b < 8)) {
        if ((a-2 >= 0 && a-2 < 8 && b+1 >= 0 && b+1 < 8) && (chessBoard[a-2][b+1]==0)) {
            counter++;
        }
        if ((a-2 >= 0 && a-2 < 8 && b-1 >= 0 && b-1 < 8) && (chessBoard[a-2][b-1]==0)) {
            counter++;
        }
        if ((a-1 >= 0 && a-1 < 8 && b+2 >= 0 && b+2 < 8) && (chessBoard[a-1][b+2]==0)) {
            counter++;
        }
        if ((a-1 >= 0 && a-1 < 8 && b-2 >= 0 && b-2 < 8) && (chessBoard[a-1][b-2]==0)) {
            counter++;
        }
        if ((a+2 >= 0 && a+2 < 8 && b+1 >= 0 && b+1 < 8) && (chessBoard[a+2][b+1]==0)) {
            counter++;
        }
        if ((a+2 >= 0 && a+2 < 8 && b-1 >= 0 && b-1 < 8) && (chessBoard[a+1][b-1]==0)) {
            counter++;
        }
        if ((a+1 >= 0 && a+1 < 8 && b+2 >= 0 && b+2 < 8) && (chessBoard[a+1][b+2]==0)) {
            counter++;
        }
        if ((a+1 >= 0 && a+1 < 8 && b-2 >= 0 && b-2 < 8) && (chessBoard[a+1][b-2]==0)) {
            counter++;
        }
    }
    if (counter < border) {
        border = counter;
        x = i;
        y = j;
    }
}

void tourBoard(int i, int j) {
    int a, b;
    a = i;
    b = j;

    if (start < 65) {
        if (chessBoard[a+2][b-1] == 0)
            if(a+2 >= 0 && a+2 < 8)
                if(b-1 >= 0 && b-1 < 8)
                    calculate(a+2, b-1);
        if(chessBoard[a+2][b+1] == 0)
            if(a+2 >= 0 && a+2 < 8)
                if(b+1 >= 0 && b+1 < 8)
                    calculate(a+2, b+1);
        if(chessBoard[a-1][b+2] == 0)
            if(a-1 >= 0 && a-1 < 8)
                if(b+2 >= 0 & b+2 < 8)
                    calculate(a-1, b+2);
        if(chessBoard[a+1][b+2] == 0)
            if(a+1 >= 0 && a+1 < 8)
                if(b+2 >= 0 && b+2 < 8)
                    calculate(a+1, b+2);
        if(chessBoard[a-2][b+1] == 0)
            if(a-2 >= 0 && a-2 < 8)
                if(b+1 >= 0 && b+1 < 8)
                    calculate(a-2, b+1);
        if(chessBoard[a-2][b-1] == 0)
            if(a-2 >= 0 && a-2 < 8)
                if(b-1 >= 0 && b-1 < 8)
                    calculate(a-2, b-1);
        if(chessBoard[a+1][b-2] == 0)
            if(a+1 >= 0 && a+1 < 8)
                if(b-2 >= 0 && b-2 < 8)
                    calculate(a+1, b-2);
        if(chessBoard[a-1][b-2] == 0)
            if(a-1 >= 0 && a-1 < 8)
                if(b-2 >= 0 && b-2 < 8)
                    calculate(a-1, b-2);

         chessBoard[i][j]=start++;
         border = 8;
         tourBoard(x, y);
    }
}

// Output
void printBoard() {
    cout << "\n Knight's Tour\n\n";
    for(int i = 0; i < 8; i++) {
          for(int j = 0; j < 8; j++)
          cout << " " << setw (2) << chessBoard[i][j];
          cout << endl << endl;
     }
     cout << endl << endl;
}
